File: src/typedb/typeql-builders.ts
/**
 * TypeQL fragments for bounded reads. Follows predicate-pushdown: constrain type + attribute
 * in the `match` before `select` (see TypeDB TypeQL optimisation guides).
 */

const SAFE_LABEL = /^[A-Za-z][A-Za-z0-9_-]*$/;

export function assertSafeTypeqlIdentifier(label: string, fieldName: string): void {
  if (!SAFE_LABEL.test(label)) {
    throw new TypeError(
      `${fieldName} must match ${SAFE_LABEL.source} to prevent TypeQL injection: got ${JSON.stringify(label)}`,
    );
  }
}

/**
 * Match entities of a concrete type that own a string attribute, then project entity + value.
 * Uses `isa` so subtypes are included (use `isa!` in schema if you need exact type only).
 */
export function buildSelectEntityWithStringAttribute(
  entityType: string,
  attributeName: string,
): string {
  assertSafeTypeqlIdentifier(entityType, "entityType");
  assertSafeTypeqlIdentifier(attributeName, "attributeName");
  return `match
  $e isa ${entityType},
  has ${attributeName} $v;
select $e, $v;
`;
}

File: app/services/typeql_builders.py
"""Safe TypeQL fragments (aligned with ``src/typedb/typeql-builders.ts``)."""

from __future__ import annotations

import re

_SAFE = re.compile(r"^[A-Za-z][A-Za-z0-9_-]*$")


def assert_safe_typeql_identifier(label: str, field_name: str) -> None:
    if not _SAFE.match(label):
        raise ValueError(
            f"{field_name} must match {_SAFE.pattern} to prevent TypeQL injection; got {label!r}",
        )


def build_select_entity_with_string_attribute(entity_type: str, attribute_name: str) -> str:
    assert_safe_typeql_identifier(entity_type, "entityType")
    assert_safe_typeql_identifier(attribute_name, "attributeName")
    return (
        "match\n"
        f"  $e isa {entity_type},\n"
        f"  has {attribute_name} $v;\n"
        "select $e, $v;\n"
    )
